Adding AppProtocol documentation for 1.18

robscott opened the pull request:
This includes updates to the Services documentation to cover the addition of a new AppProtocol field in 1.18.

Enhancement Issue: https://github.com/kubernetes/enhancements/issues/1507
PR this is documenting: https://github.com/kubernetes/kubernetes/pull/88503

/sig network

robscott commented:
Adding a hold on this until the corresponding PR is merged.

/hold

robscott commented:
Implementation PR has been merged: https://github.com/kubernetes/kubernetes/pull/88503

/hold cancel

sftim commented on content/en/docs/concepts/services-networking/service.md:
```diff
@@ -1173,6 +1173,16 @@ SCTP is not supported on Windows based nodes.
+{{< feature-state for_k8s_version="v1.18" state="alpha" >}}
+
+The AppProtocol field provides a way to specify an application protocol to be
+used for each Service port.
```
After reading this, I don't know what kind of values the field might take. Does this field it go in the `spec`? The `metadata`? Readers might not be able to work that out.

robscott replied:
Good catch, I've tried to make this more clear now, thanks!

sftim commented on content/en/docs/reference/command-line-tools-reference/feature-gates.md:
```diff
@@ -441,6 +442,7 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `ScheduleDaemonSetPods`: Enable DaemonSet Pods to be scheduled by the default scheduler instead of the DaemonSet controller.
 - `SCTPSupport`: Enables the usage of SCTP as `protocol` value in `Service`, `Endpoint`, `NetworkPolicy` and `Pod` definitions
 - `ServerSideApply`: Enables the [Sever Side Apply (SSA)](/docs/reference/using-api/api-concepts/#server-side-apply) path at the API Server.
+- `ServiceAppProtocol`: Enables the `AppProtocol` field on Services and Endpoints.
```
Does this work with EndpointSlices?

robscott replied:
EndpointSlices already have the field, so it does work with them, but it is not controlled by this feature gate.

robscott replied:
I agree with the case change suggested, but it feels strange to include EndpointSlices here. This feature gate has no direct effect on EndpointSlices. It allows Services to have `appProtocol` set, which would then get pulled into an EndpointSlice derived from that Service, but `appProtocol` can be set on an EndpointSlice regardless of the state of this feature gate.

sftim replied:
OK, that makes sense.
```suggestion
- `ServiceAppProtocol`: Enables the `appProtocol` field on Service and Endpoints objects.
```
(I haven't looked at the code for this at all).

robscott replied:
Updated, thanks!

robscott commented:
@sftim Thanks for the great feedback! 

* _to make merging easier, please move the whitespace changes to their own commit_
Done, thanks! I'd forgotten to turn off "trim trailing whitespace" on VS Code.

* _you need to document the feature. Consider providing an example manifest, or listing some example values for the field._
Added an example manifest, hope it makes sense.

* _what are the requirements for AppProtocol to work when someone sets it? Does it work with ClusterIP Services? What about ExternalName?_
There's nothing with Kubernetes to actually use or even pay attention to this field. This is primarily a hint to consumers of the API, especially load balancers, but I believe there are other use cases for this field beyond that.

* _are the permitted values of the field generic or do they depend on, say, your cloud provider?_
Added some messaging around this - they should generally use the standard IANA Service Names, but if not, a domain prefixed path is considered valid for custom app protocols. The actual API validation is fairly loose here.

* _Ideally, also add a new Task page that shows readers how enable the feature gate and then specify or change the AppProtocol for a Service to get some kind of useful outcome. The task page need not be lengthy; 100 lines of Markdown could be enough._
Do you think this is worthwhile for the limited scope of this feature? The only thing that needs to change is enabling this feature gate on kube-apiserver and kube-controller-manager, is there a standardized doc that I could reference around feature gate usage?

sftim commented on content/en/docs/concepts/services-networking/service.md:
```diff
@@ -1173,6 +1173,38 @@ SCTP is not supported on Windows based nodes.
+internal Kubernetes networking components such as kube-proxy.
+
+As an alpha feature, this field is not enabled by default. To use this field,
+enable the `ServiceAppProtocol` feature gate.
```
nit:
```suggestion
enable the `ServiceAppProtocol` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/).
```

robscott replied:
Updated, thanks!

sftim commented:
>  Do you think this is worthwhile for the limited scope of this feature? The only thing that needs to change is enabling this feature gate on kube-apiserver and kube-controller-manager, is there a standardized doc that I could reference around feature gate usage?

Now that I understand the context more, this doesn't seem like it's needed. Thanks for updating the branch / PR.

sftim commented on content/en/docs/concepts/services-networking/service.md:
```diff
@@ -1173,6 +1173,38 @@ SCTP is not supported on Windows based nodes.
 The kube-proxy does not support the management of SCTP associations when it is in userspace mode.
 {{< /warning >}}
 
+## Application protocols
```
nit / query
```suggestion
## Application protocol
```

sftim replied:
I'd consider making this a 3rd level heading that appears just before [#virtual-ips-and-service-proxies](https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies) (so, part of the [Defining a Service](https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service) section).

It's already a long page and every extra second-level heading makes the table of contents take more vertical space.

@robscott, what do you think? This point is very open to discussion.

robscott replied:
That all makes a lot of sense, updated to match that now, thanks!

sftim commented:
[Preview](https://deploy-preview-19317--kubernetes-io-vnext-staging.netlify.com/docs/concepts/services-networking/service/#application-protocols)

VineethReddy02 commented:
👋 Please rebase this PR on `dev-1.18` in order to avoid merge conflicts.

Feel free to `/hold cancel` after you've rebased.

/hold

robscott commented:
@VineethReddy02 This PR has been rebased on dev-1.18, let me know if I need to do anything else on that front.

@sftim Thanks for all the great feedback! I think I've addressed everything now.

VineethReddy02 commented:
/lgtm

jimangel commented:
/milestone 1.18

VineethReddy02 commented:
/approve
